[AArch64] Add costs for ST3 and ST4 instructions, modelled as store(shuffle).

davemgreen opened the pull request:
This tries to add some costs for the shuffle in a ST3/ST4 instruction, which are represented in LLVM IR as store(interleaving shuffle). In order to detect the store, it needs to add a CxtI context instruction to check the users of the shuffle. LD3 and LD4 are added, LD2 should be a zip1 shuffle, which will be added in another patch.

It should help fix some of the regressions from #87510.

RKSimon commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3815,18 +3815,30 @@ InstructionCost AArch64TTIImpl::getSpliceCost(VectorType *Tp, int Index) {
+             Mask, 3, Tp->getElementCount().getKnownMinValue() * 2)) &&
+        !ShuffleVectorInst::isZeroEltSplatMask(
+            Mask, Tp->getElementCount().getKnownMinValue()))
+      return LT.first;
```
Where do you use CxtI to detect a store?

davemgreen replied:
Ah, yeah! I rewrote the code when moving it into this outer if statement, and must have missed that part of the original.

davemgreen committed:
[AArch64] Add costs for ST3 and ST4 instructions, modelled as store(shuffle).

This tries to add some costs for the shuffle in a ST3/ST4 instruction, which
are represented in LLVM IR as store(interleaving shuffle). In order to detect
the store, it needs to add a CxtI context instruction to check the users of the
shuffle. LD3 and LD4 are added, LD2 should be a zip1 shuffle, which will be
added in another patch.

It should help fix some of the regressions from #87510.

RKSimon reviewed (commented):
LGTM - my only remaining thought is whether we need the Args argument now we have CxtI?

davemgreen commented:
> LGTM - my only remaining thought is whether we need the Args argument now we have CxtI?

I guess that you could know the Args that the shuffle will use without knowing/having the exact shuffle instruction already, so it might still be useful to keep.

RKSimon reviewed (approved):
OK - we can revisit this if necessary
